### PR TITLE
Padding bug in GRPO

### PR DESCRIPTION
**Pull request by krammnic**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [X] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* Fix a bug described in #2421 

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [X] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

I still need to verify that everything is correct, but here is a fix reproduction code:

```
import torch
from typing import NamedTuple, List

class GRPOTrajectory(NamedTuple):
    query_responses: torch.Tensor  # [BxG, P+L]
    logprobs: torch.Tensor         # [BxG, L]
    ref_logprobs: torch.Tensor     # [BxG, L]
    rewards: torch.Tensor          # [BxG]
    successes: torch.Tensor
    advantages: torch.Tensor       # [BxG]
    masks: torch.Tensor            # [BxG, P+L, P+L]  
    position_ids: torch.Tensor     # [BxG, P+L]
    response_padding_masks: torch.Tensor  # [BxG, L]
    seq_lens: torch.Tensor

trajectory1 = GRPOTrajectory(
    query_responses=torch.randint(0, 100, (16, 512)),
    logprobs=torch.randn(16, 512),
    ref_logprobs=torch.randn(16, 512),
    rewards=torch.randn(16),
    successes=torch.randn(16),
    advantages=torch.randn(16),
    masks=torch.ones(16, 512, 512, dtype=torch.bool),  
    position_ids=torch.arange(512).expand(16, -1),
    response_padding_masks=torch.ones(16, 512, dtype=torch.bool),
    seq_lens=torch.full((16,), 512)
)

trajectory2 = GRPOTrajectory(
    query_responses=torch.randint(0, 100, (16, 600)),
    logprobs=torch.randn(16, 300),
    ref_logprobs=torch.randn(16, 300),
    rewards=torch.randn(16),
    successes=torch.randn(16),
    advantages=torch.randn(16),
    masks=torch.ones(16, 300, 300, dtype=torch.bool), 
    position_ids=torch.arange(300).expand(16, -1),
    response_padding_masks=torch.ones(16, 300, dtype=torch.bool),
    seq_lens=torch.full((16,), 300)
)
trajectory3 = GRPOTrajectory(
    query_responses=torch.randint(0, 100, (16, 600)),
    logprobs=torch.randn(16, 300),
    ref_logprobs=torch.randn(16, 24),
    rewards=torch.randn(16),
    successes=torch.randn(16),
    advantages=torch.randn(16),
    masks=torch.ones(16, 300, 300, dtype=torch.bool),  
    position_ids=torch.arange(300).expand(16, -1),
    response_padding_masks=torch.ones(16, 300, dtype=torch.bool),
    seq_lens=torch.full((16,), 300)
)

trajectories = [trajectory1, trajectory2, trajectory3]

def _pad_tensor(tensor: torch.Tensor, target_dim: int, pad_value: float, dim: int = 1) -> torch.Tensor:
    pad_size = target_dim - tensor.shape[dim]
    if pad_size <= 0:
        return tensor
    
    pad = [0] * (2 * tensor.ndim)
    pad[-2*(dim+1)] = 0     
    pad[-2*(dim+1)+1] = pad_size
    return torch.nn.functional.pad(tensor, pad, value=pad_value)

max_p_plus_l = max(t.query_responses.shape[1] for t in trajectories)
max_l = max(t.logprobs.shape[1] for t in trajectories)

padded_trajectories = []
for traj in trajectories:
    padded_masks = _pad_tensor(
        _pad_tensor(traj.masks, max_p_plus_l, 0, dim=2),  # Pad width
        max_p_plus_l, 
        0, 
        dim=1 
    )
    
    padded_trajectories.append(GRPOTrajectory(
        query_responses=_pad_tensor(traj.query_responses, max_p_plus_l, 1, dim=1),
        logprobs=_pad_tensor(traj.logprobs, max_l, 1.0, dim=1),
        ref_logprobs=_pad_tensor(traj.ref_logprobs, max_l, 1.0, dim=1),
        rewards=traj.rewards,
        successes=traj.successes,
        advantages=traj.advantages,
        masks=padded_masks,
        position_ids=_pad_tensor(traj.position_ids, max_p_plus_l, 0, dim=1),
        response_padding_masks=_pad_tensor(traj.response_padding_masks, max_l, False, dim=1),
        seq_lens=traj.seq_lens
    ))


concatenated = GRPOTrajectory(*map(torch.cat, zip(*padded_trajectories)))
print(concatenated.query_responses.shape)
```

### CODE REVIEW

**Comment by sam-pi** [comment]
Nice! I think this is working for me, thanks

**Comment by krammnic** [comment]
Forgot about this, as we haven't iterated on a GRPO for a while. I assume that @sam-pi validated the correctness, so we can review/merge?

**Review by felipemello1 — COMMENTED** [review]
Hey Mark, thanks for the PR! we will have a new version of the recipe in torchtune soon. We can land this code. I tested it and it makes sense. But I think we would need to:

1) make the `_pad_tensor` a bit easier to understand. The indexing and arg names are a bit confusing.

for example, its not intuitive what  `pad[-2 * (dim + 1) + 1] = pad_size` is doing

Also `max_p_plus_l` only makes sense if you know in advance what p and l are.

2) think a bit about the padding values. For example, logprobs can only be <0. So prob you would want to use something like -1e9. Need to check if the other values also make sense and correctness on the loss.

If you make it a bit more intuitive, add docstrings, and verify that things still work, we can merge it.

But since we will prob have new scripts in torchtune in 2 weeks, i am not sure if its worth the effort.

**Comment by krammnic** [comment]
@felipemello1 Thanks for the review! I agree with these points.
